File: examples/mp_recovery/make_snapshot.py
#!/usr/bin/env python3
"""Create a tiny, explicitly synthetic recovery snapshot. No MP client or API key.

Formulas and provenance flags are arbitrary software fixtures, not chemistry claims.
Usage: python3 examples/mp_recovery/make_snapshot.py NEW_SNAPSHOT_DIRECTORY
"""
import argparse
import importlib.util
from pathlib import Path


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("output", type=Path)
    args = parser.parse_args()
    path = Path(__file__).resolve().parents[2] / "scripts" / "export_mp_pilot.py"
    spec = importlib.util.spec_from_file_location("mp_export", path)
    exporter = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(exporter)
    docs = []

    def add(counts, flag):
        material_id = f"mp-{len(docs) + 1}"
        docs.append(dict(material_id=material_id, composition=counts,
                         formula_pretty="synthetic", theoretical=flag,
                         database_IDs={"icsd": ["synthetic"]} if flag is False else {},
                         deprecated=False))

    for i in range(1, 11):
        add({"Li": i, "Na": 1, "O": 1}, False)
    add({"Li": 2, "Na": 2, "O": 2}, True)  # Same composition; mixed provenance.
    add({"Mg": 1, "Al": 2, "O": 4}, True)
    add({"Ca": 1, "Ti": 1, "O": 3}, True)
    add({"Ba": 1, "Ti": 1, "O": 3}, None)  # Unresolved group; never a negative.
    add({"Ba": 2, "Ti": 2, "O": 6}, True)
    exporter.write_snapshot(docs, args.output, database_version="synthetic-recovery-v1",
                            client_version="not-used-offline", is_synthetic=True)
    print(f"Created synthetic snapshot: {args.output}")


if __name__ == "__main__":
    main()

File: examples/mp_recovery/make_snapshot.jl
#!/usr/bin/env julia

"""Create a tiny, explicitly synthetic recovery snapshot without network access."""
module MakeRecoverySnapshot

using EkaCompositions

function documents()
    rows = NamedTuple[]
    function add(composition, theoretical)
        push!(rows, (material_id="mp-$(length(rows) + 1)", composition=composition,
            formula_pretty="synthetic", theoretical=theoretical,
            database_IDs=theoretical === false ? Dict("icsd" => ["synthetic"]) : Dict{String,Vector{String}}(),
            deprecated=false))
    end
    for index in 1:10
        add(Dict("Li" => index, "Na" => 1, "O" => 1), false)
    end
    add(Dict("Li" => 2, "Na" => 2, "O" => 2), true)
    add(Dict("Mg" => 1, "Al" => 2, "O" => 4), true)
    add(Dict("Ca" => 1, "Ti" => 1, "O" => 3), true)
    add(Dict("Ba" => 1, "Ti" => 1, "O" => 3), nothing)
    add(Dict("Ba" => 2, "Ti" => 2, "O" => 6), true)
    return rows
end

function main(args=ARGS)
    length(args) == 1 || throw(ArgumentError("usage: make_snapshot.jl NEW_SNAPSHOT_DIRECTORY"))
    output = write_synthetic_mp_snapshot(documents(), only(args);
        database_version="synthetic-recovery-v2")
    println("Created synthetic snapshot: $output")
    return 0
end

end

if abspath(PROGRAM_FILE) == @__FILE__
    try
        exit(MakeRecoverySnapshot.main())
    catch error
        error isa ArgumentError || rethrow()
        println(stderr, "make_snapshot: error: ", sprint(showerror, error))
        exit(2)
    end
end
